File: app/src/main/java/com/hkb48/keepdo/TaskSortingActivity.java
package com.hkb48.keepdo;

import android.content.Context;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;

import java.util.ArrayList;
import java.util.List;

public class TaskSortingActivity extends AppCompatActivity {
    private TaskAdapter mAdapter;
    private DatabaseAdapter mDBAdapter = null;
    private List<Task> mDataList = new ArrayList<>();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_task_sorting);

        final Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        assert getSupportActionBar() != null;
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mDBAdapter = DatabaseAdapter.getInstance(this);
        mDataList = mDBAdapter.getTaskList();

        mAdapter = new TaskAdapter();
        SortableListView taskListView = findViewById(R.id.mainListView);
        taskListView.setAdapter(mAdapter);
        taskListView.setDragAndDropListener(onDrop);
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
        case android.R.id.home:
            finish();
            return true;
        default:
            return super.onOptionsItemSelected(item);
        }
    }

    /**
     * Callback method for "Save" button
     */
    public void onSaveClicked(View view) {
        for (int index = 0; index < mDataList.size(); index++) {
            Task task = mDataList.get(index);
            task.setOrder(index);
            mDBAdapter.editTask(task);
        }
        finish();
    }

    /**
     * Callback method for "Cancel" button
     */
    public void onCancelClicked(View view) {
        finish();
    }

    private class TaskAdapter extends BaseAdapter {

        public int getCount() {
            return mDataList.size();
        }

        public Object getItem(int position) {
            return mDataList.get(position);
        }

        public long getItemId(int position) {
            return position;
        }

        @Override
        public boolean isEnabled(int position) {
            return false;
        }

        public View getView(int position, View convertView, ViewGroup parent) {
            final LayoutInflater inflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            View view = convertView;
            if (view == null) {
                view = inflater.inflate(R.layout.task_sorting_list_item, parent, false);
            }

            SortableListItem itemView = (SortableListItem) view;
            Task task = (Task) getItem(position);
            itemView.setText(task.getName());

            return view;
        }
    }

    private final SortableListView.DragAndDropListener onDrop = new SortableListView.DragAndDropListener() {
        public void onDrag(int from, int to) {
            if (from != to) {
                Task item = (Task) mAdapter.getItem(from);
                mDataList.remove(from);
                mDataList.add(to, item);
                mAdapter.notifyDataSetChanged();
            }
        }

        public void onDrop(int from, int to) {
            if (from != to) {
                enableSaveButton();
            }
        }
    };

    private void enableSaveButton() {
        Button okButton = findViewById(R.id.okButton);
        okButton.setEnabled(true);
    }
}

File: app/src/main/java/com/hkb48/keepdo/TaskSortingActivity.kt
package com.hkb48.keepdo

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.hkb48.keepdo.SortableListView.DragAndDropListener
import com.hkb48.keepdo.databinding.ActivityTaskSortingBinding
import com.hkb48.keepdo.databinding.TaskSortingListItemBinding
import com.hkb48.keepdo.db.entity.Task
import com.hkb48.keepdo.viewmodel.TaskViewModel
import com.hkb48.keepdo.viewmodel.TaskViewModelFactory
import com.hkb48.keepdo.widget.TasksWidgetProvider
import kotlinx.coroutines.launch


class TaskSortingActivity : AppCompatActivity() {
    private val mAdapter = TaskAdapter()
    private val taskViewModel: TaskViewModel by viewModels {
        TaskViewModelFactory(application)
    }
    private lateinit var mDataList: MutableList<Task>
    private lateinit var binding: ActivityTaskSortingBinding
    private val onDrop: DragAndDropListener = object : DragAndDropListener {
        override fun onDrag(from: Int, to: Int) {
            if (from != to) {
                val item = mAdapter.getItem(from) as Task
                mDataList.removeAt(from)
                mDataList.add(to, item)
                mAdapter.notifyDataSetChanged()
            }
        }

        override fun onDrop(from: Int, to: Int) {
            if (from != to) {
                enableSaveButton()
            }
        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityTaskSortingBinding.inflate(layoutInflater)
        setContentView(binding.root)
        val toolbar = binding.includedToolbar.toolbar
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        taskViewModel.getObservableTaskList().observe(this, { taskList ->
            taskViewModel.getObservableTaskList().removeObservers(this@TaskSortingActivity)
            mDataList = taskList.toMutableList()
            binding.mainListView.apply {
                adapter = mAdapter
                setDragAndDropListener(onDrop)
            }
            binding.cancelButton.setOnClickListener { onCancelClicked() }
            mAdapter.notifyDataSetChanged()
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (item.itemId == android.R.id.home) {
            finish()
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    /**
     * Callback method for "Save" button
     */
    private fun onSaveClicked() = lifecycleScope.launch {
        for (index in mDataList.indices) {
            val task = mDataList[index]
            val newTask = Task(
                task._id,
                task.name,
                task.monFrequency,
                task.tueFrequency,
                task.wedFrequency,
                task.thrFrequency,
                task.friFrequency,
                task.satFrequency,
                task.sunFrequency,
                task.context,
                task.reminderEnabled,
                task.reminderTime,
                index
            )
            taskViewModel.editTask(newTask)
        }
        TasksWidgetProvider.notifyDatasetChanged(applicationContext)
        finish()
    }

    /**
     * Callback method for "Cancel" button
     */
    private fun onCancelClicked() {
        finish()
    }

    private fun enableSaveButton() {
        binding.okButton.apply {
            isEnabled = true
            setOnClickListener { onSaveClicked() }
        }
    }

    private inner class TaskAdapter : BaseAdapter() {
        override fun getCount(): Int {
            return mDataList.size
        }

        override fun getItem(position: Int): Any {
            return mDataList[position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun isEnabled(position: Int): Boolean {
            return false
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view = convertView ?: TaskSortingListItemBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ).root
            val itemView = view as SortableListItem
            val task = getItem(position) as Task
            itemView.setText(task.name)
            return view
        }
    }
}